File: src/hooks/ir.rs
use std::collections::BTreeMap;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use serde_json::Value;

// Re-exported so existing `crate::hooks::ir::HarnessTarget` paths (CLI args, spec field) resolve
// after `HookOutputTarget` was merged into the canonical `HarnessTarget`.
pub(crate) use crate::artifacts::HarnessTarget;

#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
pub enum ClaudeEvent {
    PreToolUse,
    PostToolUse,
    UserPromptSubmit,
    Stop,
    SubagentStop,
    SessionStart,
    SessionEnd,
    PreCompact,
    Notification,
    PermissionRequest,
}

impl ClaudeEvent {
    pub fn as_claude_str(self) -> &'static str {
        match self {
            Self::PreToolUse => "PreToolUse",
            Self::PostToolUse => "PostToolUse",
            Self::UserPromptSubmit => "UserPromptSubmit",
            Self::Stop => "Stop",
            Self::SubagentStop => "SubagentStop",
            Self::SessionStart => "SessionStart",
            Self::SessionEnd => "SessionEnd",
            Self::PreCompact => "PreCompact",
            Self::Notification => "Notification",
            Self::PermissionRequest => "PermissionRequest",
        }
    }

    /// Codex's hook schema (codex-rs/hooks) rejects unknown keys silently via serde,
    /// so names must match its PascalCase enum variants exactly. Codex currently only
    /// handles `PreToolUse`, `PostToolUse`, `SessionStart`, `UserPromptSubmit`, `Stop`;
    /// other events emit here but Codex will ignore them, which is fine.
    pub fn as_codex_str(self) -> &'static str {
        self.as_claude_str()
    }

    pub fn from_any_str(value: &str) -> Option<Self> {
        match value {
            "PreToolUse" | "pre-tool-use" => Some(Self::PreToolUse),
            "PostToolUse" | "post-tool-use" => Some(Self::PostToolUse),
            "UserPromptSubmit" | "user-prompt-submit" => Some(Self::UserPromptSubmit),
            "Stop" | "stop" => Some(Self::Stop),
            "SubagentStop" | "subagent-stop" => Some(Self::SubagentStop),
            "SessionStart" | "session-start" => Some(Self::SessionStart),
            "SessionEnd" | "session-end" => Some(Self::SessionEnd),
            "PreCompact" | "pre-compact" => Some(Self::PreCompact),
            "Notification" | "notification" => Some(Self::Notification),
            "PermissionRequest" | "permission-request" => Some(Self::PermissionRequest),
            _ => None,
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
pub enum HookLayer {
    SeededNative,
    PackPlugin,
    BareSkill,
    DotAgents,
}

impl HookLayer {
    pub fn sort_rank(self) -> u8 {
        match self {
            Self::SeededNative => 0,
            Self::PackPlugin => 1,
            Self::BareSkill => 2,
            Self::DotAgents => 3,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct CommandHandler {
    pub command: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timeout_secs: Option<u64>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct HttpHandler {
    pub url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub method: Option<String>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub headers: BTreeMap<String, String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub body: Option<Value>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PromptHandler {
    pub prompt: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct AgentHandler {
    pub prompt: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub agent: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum ClaudeHandler {
    Command(CommandHandler),
    Http(HttpHandler),
    Prompt(PromptHandler),
    Agent(AgentHandler),
}

impl ClaudeHandler {
    pub fn kind_name(&self) -> &'static str {
        match self {
            Self::Command(_) => "command",
            Self::Http(_) => "http",
            Self::Prompt(_) => "prompt",
            Self::Agent(_) => "agent",
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct HookOrigin {
    pub layer: HookLayer,
    pub module: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cache_key: Option<String>,
    pub source_rel: String,
    pub source_root: PathBuf,
    pub source_file: PathBuf,
    pub package_key: String,
    pub event_index: usize,
    pub matcher_group_index: usize,
    pub hook_index: usize,
}

impl HookOrigin {
    pub fn source_id(&self) -> String {
        format!(
            "{}:{}:{}:{}",
            self.layer.sort_rank(),
            self.module,
            self.package_key,
            self.source_rel
        )
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct NormalizedHook {
    pub event: ClaudeEvent,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub matcher: Option<String>,
    pub handler: ClaudeHandler,
    pub origin: HookOrigin,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub matcher_group_extra: BTreeMap<String, Value>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub raw_extra: BTreeMap<String, Value>,
}

impl NormalizedHook {
    pub fn is_strict(&self) -> bool {
        matches!(
            self.event,
            ClaudeEvent::PreToolUse | ClaudeEvent::PermissionRequest
        )
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct HookBundle {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub hooks: Vec<NormalizedHook>,
}

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub enum HookDecision {
    #[default]
    Allow,
    Ask,
    Deny,
}

impl HookDecision {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Allow => "allow",
            Self::Ask => "ask",
            Self::Deny => "deny",
        }
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct NormalizedHookResult {
    pub decision: HookDecision,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub additional_context: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub updated_input: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub updated_tool_output: Option<Value>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub metadata: BTreeMap<String, Value>,
}

File: internal/hooks/ir.go
package hooks

import (
	"encoding/json"
	"fmt"
)

type Event string

const (
	PreToolUse        Event = "PreToolUse"
	PostToolUse       Event = "PostToolUse"
	UserPromptSubmit  Event = "UserPromptSubmit"
	Stop              Event = "Stop"
	SubagentStop      Event = "SubagentStop"
	SessionStart      Event = "SessionStart"
	SessionEnd        Event = "SessionEnd"
	PreCompact        Event = "PreCompact"
	Notification      Event = "Notification"
	PermissionRequest Event = "PermissionRequest"
)

func ParseEvent(value string) (Event, bool) {
	aliases := map[string]Event{
		"PreToolUse": PreToolUse, "pre-tool-use": PreToolUse,
		"PostToolUse": PostToolUse, "post-tool-use": PostToolUse,
		"UserPromptSubmit": UserPromptSubmit, "user-prompt-submit": UserPromptSubmit,
		"Stop": Stop, "stop": Stop, "SubagentStop": SubagentStop, "subagent-stop": SubagentStop,
		"SessionStart": SessionStart, "session-start": SessionStart,
		"SessionEnd": SessionEnd, "session-end": SessionEnd,
		"PreCompact": PreCompact, "pre-compact": PreCompact,
		"Notification": Notification, "notification": Notification,
		"PermissionRequest": PermissionRequest, "permission-request": PermissionRequest,
	}
	event, found := aliases[value]
	return event, found
}

type Layer string

const (
	SeededNative Layer = "SeededNative"
	PackPlugin   Layer = "PackPlugin"
	BareSkill    Layer = "BareSkill"
	DotAgents    Layer = "DotAgents"
)

func (layer Layer) Rank() int {
	switch layer {
	case SeededNative:
		return 0
	case PackPlugin:
		return 1
	case BareSkill:
		return 2
	default:
		return 3
	}
}

type HandlerKind string

const (
	CommandHandler HandlerKind = "command"
	HTTPHandler    HandlerKind = "http"
	PromptHandler  HandlerKind = "prompt"
	AgentHandler   HandlerKind = "agent"
)

type Handler struct {
	Kind    HandlerKind
	Command string
	Timeout *uint64
	URL     string
	Method  string
	Headers map[string]string
	Body    any
	Prompt  string
	Agent   string
	Model   string
}

// MarshalJSON preserves the Rust serde externally-tagged representation used
// by hook execution specs.
func (handler Handler) MarshalJSON() ([]byte, error) {
	inner := make(map[string]any)
	tag := ""
	switch handler.Kind {
	case CommandHandler:
		tag, inner["command"] = "Command", handler.Command
		if handler.Timeout != nil {
			inner["timeout_secs"] = *handler.Timeout
		}
	case HTTPHandler:
		tag, inner["url"] = "Http", handler.URL
		if handler.Method != "" {
			inner["method"] = handler.Method
		}
		if len(handler.Headers) != 0 {
			inner["headers"] = handler.Headers
		}
		if handler.Body != nil {
			inner["body"] = handler.Body
		}
	case PromptHandler:
		tag, inner["prompt"] = "Prompt", handler.Prompt
		if handler.Model != "" {
			inner["model"] = handler.Model
		}
	case AgentHandler:
		tag, inner["prompt"] = "Agent", handler.Prompt
		if handler.Agent != "" {
			inner["agent"] = handler.Agent
		}
		if handler.Model != "" {
			inner["model"] = handler.Model
		}
	default:
		return nil, fmt.Errorf("unknown handler kind %q", handler.Kind)
	}
	return json.Marshal(map[string]any{tag: inner})
}

func (handler *Handler) UnmarshalJSON(data []byte) error {
	var outer map[string]json.RawMessage
	if err := json.Unmarshal(data, &outer); err != nil {
		return err
	}
	if len(outer) != 1 {
		return fmt.Errorf("hook handler must contain exactly one variant")
	}
	for tag, raw := range outer {
		var inner struct {
			Command     string            `json:"command"`
			TimeoutSecs *uint64           `json:"timeout_secs"`
			URL         string            `json:"url"`
			Method      string            `json:"method"`
			Headers     map[string]string `json:"headers"`
			Body        any               `json:"body"`
			Prompt      string            `json:"prompt"`
			Agent       string            `json:"agent"`
			Model       string            `json:"model"`
		}
		if err := json.Unmarshal(raw, &inner); err != nil {
			return err
		}
		switch tag {
		case "Command":
			handler.Kind = CommandHandler
		case "Http":
			handler.Kind = HTTPHandler
		case "Prompt":
			handler.Kind = PromptHandler
		case "Agent":
			handler.Kind = AgentHandler
		default:
			return fmt.Errorf("unknown hook handler variant %q", tag)
		}
		handler.Command, handler.Timeout = inner.Command, inner.TimeoutSecs
		handler.URL, handler.Method, handler.Headers, handler.Body = inner.URL, inner.Method, inner.Headers, inner.Body
		handler.Prompt, handler.Agent, handler.Model = inner.Prompt, inner.Agent, inner.Model
	}
	return nil
}

type Origin struct {
	Layer             Layer  `json:"layer"`
	Module            string `json:"module"`
	CacheKey          string `json:"cache_key,omitempty"`
	SourceRelative    string `json:"source_rel"`
	SourceRoot        string `json:"source_root"`
	SourceFile        string `json:"source_file"`
	PackageKey        string `json:"package_key"`
	EventIndex        int    `json:"event_index"`
	MatcherGroupIndex int    `json:"matcher_group_index"`
	HookIndex         int    `json:"hook_index"`
}

func (origin Origin) SourceID() string {
	return fmt.Sprintf("%d:%s:%s:%s", origin.Layer.Rank(), origin.Module, origin.PackageKey, origin.SourceRelative)
}

type Hook struct {
	Event             Event          `json:"event"`
	Matcher           string         `json:"matcher,omitempty"`
	Handler           Handler        `json:"handler"`
	Origin            Origin         `json:"origin"`
	MatcherGroupExtra map[string]any `json:"matcher_group_extra,omitempty"`
	RawExtra          map[string]any `json:"raw_extra,omitempty"`
}

func (hook Hook) Strict() bool { return hook.Event == PreToolUse || hook.Event == PermissionRequest }

type Bundle struct {
	Hooks []Hook `json:"hooks,omitempty"`
}

type Decision string

const (
	Allow Decision = "allow"
	Ask   Decision = "ask"
	Deny  Decision = "deny"
)

type Result struct {
	Decision          Decision       `json:"decision"`
	Message           string         `json:"message,omitempty"`
	AdditionalContext string         `json:"additional_context,omitempty"`
	UpdatedInput      any            `json:"updated_input,omitempty"`
	UpdatedToolOutput any            `json:"updated_tool_output,omitempty"`
	Metadata          map[string]any `json:"metadata,omitempty"`
}
